crypto/evp/pmeth_lib.c

```diff
@@ -1665,8 +1677,33 @@ static int evp_pkey_ctx_store_cached_data(EVP_PKEY_CTX *ctx,
                                           int cmd, const char *name,
                                           const void *data, size_t data_len)
 {
-    if ((keytype != -1 && ctx->pmeth->pkey_id != keytype)
-        || ((optype != -1) && !(ctx->operation & optype))) {
+    if (keytype != -1) {
+        switch (evp_pkey_ctx_state(ctx)) {
+        case EVP_PKEY_STATE_PROVIDER:
+            if (ctx->keymgmt == NULL) {
+                ERR_raise(ERR_LIB_EVP, EVP_R_COMMAND_NOT_SUPPORTED);
+                return -2;
+            }
+            if (!EVP_KEYMGMT_is_a(ctx->keymgmt,
+                                  evp_pkey_type2name(keytype))) {
```

Comment:
For keytype = EVP_PKEY_DHX the name used would be "X9.42 DH"
Does this alias work with EVP_KEYMGMT_is_a() for "DHX"?

Reply:
IMO yes. Some tests and some functionality already depend on this.

Reply:
https://github.com/openssl/openssl/blob/a2a5506b9329b978a2a5b11a518b9789446ad310/providers/defltprov.c#L412-L413
https://github.com/openssl/openssl/blob/a2a5506b9329b978a2a5b11a518b9789446ad310/providers/fips/fipsprov.c#L398-L399